oci/archive/oci_src.go

```diff
@@ -0,0 +1,153 @@
+		}
+		return imgspecv1.Descriptor{}, errors.Wrap(err, "error loading manifest")
+	}
+	return descriptor, nil
+}
+
+// Reference returns the reference used to set up this source.
+func (s *ociArchiveImageSource) Reference() types.ImageReference {
+	return s.ref
+}
+
+// Close removes resources associated with an initialized ImageSource, if any.
+// Close deletes the temporary directory at dst
+func (s *ociArchiveImageSource) Close() error {
+	if err := s.unpackedSrc.Close(); err != nil {
+		if err := deleteTempDir(); err != nil {
+			return errors.Wrapf(err, "error deleting temp directory", tempDirRef.tempDirectory)
+		}
+		return errors.Wrapf(err, "error closing image source")
+	}
+	if err := deleteTempDir(); err != nil {
+		return errors.Wrapf(err, "error deleting temp directory", tempDirRef.tempDirectory)
+	}
+	return nil
+}
+
+// GetManifest returns the image's manifest along with its MIME type
+// (which may be empty when it can't be determined but the manifest is available).
+func (s *ociArchiveImageSource) GetManifest() ([]byte, string, error) {
+	return s.unpackedSrc.GetManifest()
+}
+
+func (s *ociArchiveImageSource) GetTargetManifest(digest digest.Digest) ([]byte, string, error) {
+	valByte, valString, err := s.unpackedSrc.GetTargetManifest(digest)
+	if err != nil {
+		if err := deleteTempDir(); err != nil {
+			return nil, "", errors.Wrapf(err, "error deleting temp directory", tempDirRef.tempDirectory)
+		}
+		return nil, "", err
+	}
+	return valByte, valString, nil
+}
+
+// GetBlob returns a stream for the specified blob, and the blob's size.
+func (s *ociArchiveImageSource) GetBlob(info types.BlobInfo) (io.ReadCloser, int64, error) {
+	ioReader, valInt, err := s.unpackedSrc.GetBlob(info)
+	if err != nil {
+		if err := deleteTempDir(); err != nil {
+			return nil, 0, errors.Wrapf(err, "error deleting temp directory", tempDirRef.tempDirectory)
+		}
+		return nil, 0, err
+	}
+	return ioReader, valInt, nil
+}
+
+func (s *ociArchiveImageSource) GetSignatures(c context.Context) ([][]byte, error) {
+	valByte, err := s.unpackedSrc.GetSignatures(c)
+	if err != nil {
+		if err := deleteTempDir(); err != nil {
+			return nil, errors.Wrapf(err, "error deleting temp directory", tempDirRef.tempDirectory)
+		}
+		return nil, err
+	}
+	return valByte, nil
+}
+
+// unTar unpacks the tar file at src and saves it at dst
+func unTar(src, dst string) error {
+	if err := archive.UntarPath(src, dst); err != nil {
+		return errors.Wrapf(err, "error untarring file %q", src)
+	}
+	return nil
+}
+
+// deleteTemDir deletes the temporary directory created for the oci image
+func deleteTempDir() error {
+	if tempDirRef != nil {
+		return os.RemoveAll(tempDirRef.tempDirectory)
+	}
+	return nil
+}
+
+// creates the temporary directory and copies the tarred content to it
+func createUntarTempDir(ref ociArchiveReference) error {
+	var err error
+	tempDirRef, err = createOciRef(ref.image)
```

Comment:
Non-blocking nit: `createOciRef` is not quite explicit about what the method modifying the filesystem, and `createUntarTempDir` with two verbs is a tiny bit difficult to follow (and untar what?). Maybe something like `[create]temp[orary]OCILayout[Dir]`, `extract[ed]OCILayout[FromArchive]`? I guess that’s verging towards excessive verbosity